### PR TITLE
allow trailling comma in decl_event

### LINKED ISSUE #2114: `decl_event` cannot handle trailing comma
With this code:

```
decl_event!(
	pub enum Event<T>
	where
		Amount = AssetOf<T>
	{
		/// Fee charged (extrinsic_index, fee_amount)
		Charged(u32, Amount),
	}
);
```

rustfmt attempts to insert trailing comma

```
 decl_event!(
    pub enum Event<T>
    where
-       Amount = AssetOf<T>
+       Amount = AssetOf<T>,
    {
        /// Fee charged (extrinsic_index, fee_amount)
        Charged(u32, Amount),
```

which breaks `decl_event`

### PR DESCRIPTION
**Pull request by xlc**
Fixes #2114 

Mostly just change the order of parse and finish arm. And make the comma optional in the finish arm.

### CODE REVIEW

**Comment by gavofyork** [comment]
needs a runtime `impl_version` bump

**Comment by bkchr** [comment]
Do we really need to bump the implementation version for this? This just changes a macro without changing any generated code.

File: srml/support/src/event.rs [hunk 1]
```diff
@@ -183,79 +183,79 @@ macro_rules! __decl_generic_event {
...
 	) => {
-		$crate::__decl_generic_event!(@format_generic
+		$crate::__decl_generic_event!(@generate
 			$( #[ $attr ] )*;
 			$event_generic_param;
 			$($instance $( = $event_default_instance)? )?;
-			{ $($rest)* };
-			{ $($parsed)*, $generic_rename = $generic_type };
+			{ $($events)* };
+			{ $($parsed)*, $trait_type = <$generic as $trait>::$trait_type};
```
Comment (bkchr) [inline srml/support/src/event.rs]:
```suggestion
			{ $($parsed)*, $trait_type = <$generic as $trait>::$trait_type };
```

File: srml/support/src/event.rs [hunk 2]
```diff
@@ -183,79 +183,79 @@ macro_rules! __decl_generic_event {
...
-	// Parse unnamed
+	// Finish formatting on a named one
 	(@format_generic
 		$(#[$attr:meta])*;
 		$event_generic_param:ident;
 		$($instance:ident $( = $event_default_instance:path)? )?;
-		{ <$generic:ident as $trait:path>::$trait_type:ident, $($rest:tt)* };
-		{$($parsed:tt)*};
+		{ $generic_rename:ident = $generic_type:ty $(,)? { $( $events:tt )* } };
+		{$( $parsed:tt)*};
```
Comment (bkchr) [inline srml/support/src/event.rs]:
```suggestion
		{ $($parsed:tt)* };
```

File: srml/support/src/event.rs [hunk 3]
```diff
@@ -183,79 +183,79 @@ macro_rules! __decl_generic_event {
...
-		$crate::__decl_generic_event!(@format_generic
-			$( #[ $attr ] )*;
+		$crate::__decl_generic_event!(@generate
+			$(#[$attr])*;
 			$event_generic_param;
 			$($instance $( = $event_default_instance)? )?;
-			{ $($rest)* };
-			{ $($parsed)*, $trait_type = <$generic as $trait>::$trait_type };
+			{ $($events)* };
+			{ $($parsed)*, $generic_rename = $generic_type};
```
Comment (bkchr) [inline srml/support/src/event.rs]:
```suggestion
			{ $($parsed)*, $generic_rename = $generic_type };
```

File: srml/support/src/event.rs [hunk 4]
```diff
@@ -183,79 +183,79 @@ macro_rules! __decl_generic_event {
...
-	// Finish formatting on a named one
+	// Unnamed type can't be parsed
 	(@format_generic
 		$(#[$attr:meta])*;
 		$event_generic_param:ident;
 		$($instance:ident $( = $event_default_instance:path)? )?;
-		{ $generic_rename:ident = $generic_type:ty { $( $events:tt )* } };
-		{$( $parsed:tt)*};
+		{ $generic_type:ty, $($rest:tt)* };
+		{$($parsed:tt)*};
```
Comment (bkchr) [inline srml/support/src/event.rs]:
```suggestion
		{ $($parsed:tt)* };
```

File: srml/support/src/event.rs [hunk 5]
```diff
@@ -183,79 +183,79 @@ macro_rules! __decl_generic_event {
...
-		{ $generic_type:ty, $($rest:tt)* };
-		{$($parsed:tt)*};
+		{ $generic_rename:ident = $generic_type:ty, $($rest:tt)* };
+		{$( $parsed:tt)*};
 	) => {
-		$crate::__decl_generic_event!(@cannot_parse $generic_type);
+		$crate::__decl_generic_event!(@format_generic
+			$( #[ $attr ] )*;
+			$event_generic_param;
+			$($instance $( = $event_default_instance)? )?;
```
Comment (bkchr) [inline srml/support/src/event.rs]:
```suggestion
			$( $instance $( = $event_default_instance)? )?;
```

**Pull request merged**